File: 프로그래머스/0/120869. 외계어 사전/외계어 사전.js
function solution(spell, dic) {
    const visited = Array(spell.length).fill(false);
    let found = false;
    
    function permute(arr) {
        if (arr.length === spell.length) {
            if (dic.includes(arr.join(""))) {
                found = true;
            }
            return;
        }
        
        for (let i = 0; i < spell.length; i++) {
            if (visited[i]) continue;
            visited[i] = true;
            arr.push(spell[i]);
            permute(arr);
            if (found) return;
            arr.pop();
            visited[i] = false;
        }

    }
    
    permute([]);
    
    return found ? 1 : 2;
}

File: 프로그래머스/0/120869. 외계어 사전/외계어 사전.py
def solution(spell, dic):
    arr = []
    found = [False]
    visited = [False] * len(spell)
    
    def permute():
        if len(arr) == len(spell):
            word = ''.join(arr)
            if word in dic:
                found[0] = True
            return
        for i in range(len(spell)):
            if visited[i]:
                continue
            arr.append(spell[i])
            visited[i] = True
            permute()
            arr.pop()
            visited[i] = False
            
    permute()         
    return 1 if found[0] else 2